### ISSUE TITLE
Ignore unused code

### ISSUE THREAD

**Issue by Remo** [peachpiecompiler/peachpie#289]
I'm trying to get some of my projects working on peachpie. Two projects have pretty much the same issue. When I try to execute `dotnet run`, I get errors like this:
```
C:\work\eloquent-peachpie\vendor\symfony\translation\Tests\TranslatorCacheTest.php(21,35): error PHP4044: Type name 'PHPUnit\Framework\TestCase' could not be resolved. [C:\work\eloquent-peachpie\eloquent-peachpie.msbuildproj]
```

That's correct, phpunit doesn't exist in my project and I don't want to include it, but some libraries I use contain unit tests based on phpunit. I tried to modify my msbuild file like this:
```
<Compile Include="**/*.php" Exclude="**/phpunit**;**/Tests/**;**/Seeds/**;**/Testing/**;**/Migrations/**;**/TranslatorPass.php;**/TranslationExtractorPass.php;**/TranslationDataCollector.php;**/TranslationDumperPass.php;**/FactoryMakeCommand.php;**/XliffLintCommand.php"/>
```

This helped me to get past those errors, but it's a bit tedious. Please note that simply excluding the tests wasn't enough.

Long story short: Is there a way to include/compile only the files which are actually used? With composer dependencies we often have a ton of code which is never used.

**Comment by jakubmisek** [comment]
That's a good point. Currently the `Compile`'s `Exclude` is everything we have.

I'm currently thinking of two solutions:
- Add a dummy PHP script containing mock `PHPUnit\Framework\TestCase`. Then the compilation would succeed.
- PeachPie would have an option for ignoring declarations based on certain conditions (e.g. it depends on PHPUnit\Framework\TestCase).

**Comment by Remo** [comment]
I'm hesitant about the PHP mock. It might work for PHPUnit but those kind of things will keep on showing  up. I guess it's possible to resolve the code starting from the entry point? Kind of like tree shaking to remove dead (unused) code.

**Comment by Remo** [comment]
Yes, I see, I guess PHP is a bit too dynamic there. Maybe if we supported some kind of framework templates it wouldn't be much of an issue. At some point when symfony works we could create a symfony template that excludes all the files not needed to run a symfony project. That way it would be easier for people to get started. Maybe something for the future. Or we'll wait for some stackoverflow articles, that should help too ;-) I was just surprised about the long list of errors, but it all makes sense.

**Comment by liesauer** [comment]
```
<Compile Include="**/*.php" />
<Compile Remove="**/test/**/*.php" />
<Compile Remove="**/tests/**/*.php" />
```
this will be a little better. i have been use it for a long time.

**Comment by Remo** [comment]
@liesauer thanks, that's easier to read, didn't even try that. Saw the semicolon approach and went for it, but will use yours.

**Comment by liesauer** [comment]
btw, do not rashly include any composer libraries, it will make tons of warning and errors.

**Comment by Remo** [comment]
@liesauer I'm aware of that, but how would you start if you wanted to get something like laravel running? I'm a big fan of composer in the world of PHP and all projects I work with, use composer. I could fork every library and start picking the files I need, but that's worse than excluding some files.

**Comment by liesauer** [comment]
if the warnings and errors are made by the library's tests, just ignore it. but if made by the library's core  codes, there is no way out up to now, a little library may be dependent on a lot of dependencies, and cause chain-reacting.

**Comment by jakubmisek** [comment]
closing as no applicable
